app/controllers/application_controller.rb

```diff
@@ -1,10 +1,154 @@
+
+  get '/signup' do
+    if logged_in?
+      redirect to '/tweets'
+    else
+      erb :'users/create_user'
+    end
+  end
+
+  post '/signup' do
+    if params[:username] == "" || params[:email] == "" || params[:password] == ""
+      redirect to '/signup'  #how can you raise an error here? "You must fill in the all parts of the form to create an account."
+    else
+      @user = User.create(username: params[:username], email: params[:email], password: params[:password])
+      @user.save
+      session[:user_id] = @user.id
+      redirect to '/tweets'
+    end
+  end
+
+  get '/login' do
+    if logged_in?
+      redirect to '/tweets'
+    else
+      erb :'users/login'
+    end
+  end
+
+  post '/login' do
+    @user = User.find_by(username: params[:username])
+    if @user && @user.authenticate(params[:password])
+      session[:user_id] = @user.id
+      redirect '/tweets'
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/logout' do
+    if logged_in?
+      session.clear
+      redirect to '/login'
+    else
+      redirect to '/'
+    end
+  end
+
+  get '/users/:slug' do  #get "/users/#{user.slug}" shows all a single users tweets'
+    @user = User.find_by_slug(params[:slug])
+    erb :'users/show'
+  end
+
+  get '/tweets' do
+    if logged_in?
+      @user = User.find_by_id(session[:user_id])
+      @tweets = Tweet.all
+      erb :'tweets/tweets'
+    else
+      redirect to '/login'
+    end
+  end
+
+  get '/tweets/new' do
+    if logged_in?
+      erb :'tweets/create_tweet'
+    else
+      redirect to '/login'
+    end
+  end
+
+  post '/tweets' do
+    if params[:content] != ""
+      @tweet = Tweet.create(content: params[:content], user_id: session[:user_id])
+      redirect to  "/tweets/#{@tweet.id}"
+    else
+      redirect to '/tweets/new'
+    end
+  end
+
+  get '/tweets/:id' do
+    if logged_in?
+      @tweet = Tweet.find_by_id(params[:id])
+      erb :'tweets/show_tweet'
+    else
+      redirect to '/login'
+    end
+  end
+
+  get '/tweets/:id/edit' do
+    if logged_in?
+      @tweet = Tweet.find_by_id(params[:id])
+      erb :'/tweets/edit_tweet'
+    else
+      redirect to '/login'
+    end
+  end
+
+  patch '/tweets/:id' do
+    if logged_in?
+      @tweet = Tweet.find_by_id(params[:id])
+      if params[:content] == ""
+        redirect to "/tweets/#{@tweet.id}/edit"
+      else
+        @tweet.content = params[:content]
+        @tweet.save
+        redirect to  "/tweets/#{@tweet.id}"
+      end
+    else
+      redirect to '/login'
+    end
+  end
+
+  delete '/tweets/:id/delete' do
+    if logged_in?
+      @tweet = Tweet.find_by_id(params[:id])
+      if @tweet.user_id == session[:user_id]
+        @tweet.delete
+        redirect to '/tweets'
+      else
+        redirect to '/tweets'
+      end
+    else
+      redirect to '/login'
+    end
+  end
+
+
+  helpers do
+    def logged_in?
+      !!current_user
+    end
+
+    def current_user
+      @current_user = session.has_key?(:user_id)
```

Comment:
This method should find a user based on the user_id key in the session hash. This is a proper way to implement the method. 

```ruby
def current_user
  User.find(session[:id]) if session[:id]
end
```

Because in the lifespan of a request, you might call `current_user` multiple times, we'd prefer this method to only hit the database once and then 'cache' the user instance into an instance variable that could be returned on any subsequent call to `current_user`.

```ruby
def current_user
  @current_user ||= User.find(session[:id]) if session[:id]
end
```

This pattern of conditionally setting an instance variable or returning it if it's already defined using the `||=` (or-equals) operator is called Memoization ([read more](http://gavinmiller.io/2013/basics-of-ruby-memoization/) and [here](http://www.justinweiss.com/articles/4-simple-memoization-patterns-in-ruby-and-one-gem/)).